ABCI++ spec. First complete version (#366)
tendermint/spec: spec/abci++/abci++_properties_001_draft.md

---
order: 1
title: New Methods
---

# New Methods

## Description

### PrepareProposal

#### Parameters and Return values

TODO

#### When does Tendermint call it?

When a validator _p_ enters Tendermint consensus round _r_, height _h_, in which _p_ is the proposer:

> 1. _p_'s Tendermint collects the outstanding transactions from the mempool (TODO: should we limit size & gas?).

Reviewer: ```suggestion
1. _p_'s Tendermint collects the outstanding transactions from the mempool (TODO: should we limit size & gas?).
```
```suggestion
1. _p_'s Tendermint collects outstanding transactions from the mempool (TODO: should we limit size & gas?).
```

2. _p_'s Tendermint creates a block header.
3. _p_'s Tendermint calls `PrepareProposal` with the newly created block. The call is synchronous (i.e., Tendermint's execution will block until the Application returns).
4. The Application checks the block (the header and transactions). It can also:
   * add/remove transactions
   * modify the header hashes
   * if the block is modified, the Application sets it in the return parameters

> 5. The Application signals _Accept_ of _Reject_ in `PrepareProposal`'s return values

Reviewer: ```suggestion
5. The Application signals _Accept_ or _Reject_ in `PrepareProposal`'s return values
```

Reviewer: Does this mean that a bug in the application ("always reject", or "reject for three weeks") may hold the chain?

   * TODO: Decide if this kind if Accept/Reject is wanted/covered by impl (maybe a panic here makes more sense?)
   * If _Reject_, the proposed block --along with any modification-- is discarded. Tendermint interprets that there is nothing to be proposed for consensus at the moment and _p_ won't send any proposal message in round _r_, height _h_.
   * If _Accept_, _p_'s Tendermint uses the modified block as _p_'s proposal in round _r_, height _h_.

### ProcessProposal

#### Parameters and Return values

TODO

#### When does Tendermint call it?

When a validator _p_ enters Tendermint consensus round _r_, height _h_, in which _q_ is the proposer (possibly _p_ = _q_):

1. _p_ sets up timer `ProposeTimeout`.
2. If _p_ is the proposer, execute steps 1-6 in _PrepareProposal_.

> 3. Upon reception of `Proposal` message from _q_ for round _r_, height _h_, _p_'s Tendermint calls `ProcessProposal` with the newly received proposal. The call is synchronous.

Reviewer: Is the the proposal or the proposal and the block?

Author: I would say both, but let's discuss it later

Author: Discussed: both. Text updated

4. The Application checks/processes the proposed block, which is read-only, and returns _Accept_ or _Reject_.
   * Depending on the Application's needs, it may return from `ProcessProposal`
     * either after it has completely processed the block (the simpler case),
     * or immediately (after doing some basic checks), and process the block asynchronously. In this case the Application won't be able to reject the block, or force prevote/precommit `nil` afterwards.
5. If the returned value is
     * _Reject_, Tendermint will prevote `nil` for the proposal in round _r_, height _h_ (this can be interpreted as _valid(v)_ returning false in the white paper's pseudocode)
     * _Accept_, Tendermint will follow the normal algorithm to prevote on the proposal in _r_, height _h_

### ExtendVote

#### Parameters and Return values

TODO

#### When does Tendermint call it?

When a validator _p_ is in Tendermint consensus round _r_, height _h_, state _prevote_, and the other conditions prescribed by the consensus algorithm for sending a _Precommit_ message are fulfilled:

> 1. _p_'s Terndermint updates the consensus state as prescriped by the algorithm (_validValue_, _lockedValue_, etc.)

Reviewer: "consensus state" is a term that is (mis)used in IBC for a block header. We should figure out to call the state here...

Author: Yes, the wording of the highlighted text is "shallow" (I mean, ambiguous). I left it this way until we discuss extending votes for `nil` precommits, since that will affect the final wording of this text.

Author: Discussion: agreed to reword this section assuming we won't be extending `nil` precommits

Author: Replaced consensus state by "_lock*_ and _valid*_ variables"

2. _p_'s Tendermint calls `ExtendVote` with the value that is about to be sent as precommit message. The call is synchronous.
3. The Application returns an array of bytes, `precommit_extension`, which is not interpreted by Tendermint.
4. _p_'s Tendermint includes `precommit_extension` as a new field in the `precommit` message.

> 5. _p_'s Tendermint broadcasts the `precommit` message.

Reviewer: Can it happen that a validator is byzantine and equivocates? Perhaps only on the `precommit_extension`?

Author: Indeed. That is something we need to explore IMO. But first, I think I need a discussion with Dev (and others?) to understand how exactly they intend to use this field.

Author: Discussion: we came up with a set of rules for ExtendVote, which will be included in the next commit to this PR

---
order: 1
title: New Methods
---

# New Methods

## Description

### PrepareProposal

#### Parameters and Types

>**TODO**: Hyperlinks for ConsensusParams, LastCommitInfo, Evidence, Event, and ValidatorUpdate are
broken because they are defined in abci.md (and not copied over for the moment).

>**TODO**: don't forget to extend the Header structure with "bool problem" when including the parts
common with the existing ABCI spec.

* **Request**:

    | Name                    | Type                                        | Description                                                                                                | Field Number |
    |-------------------------|---------------------------------------------|------------------------------------------------------------------------------------------------------------|--------------|

>     | hash                    | bytes                                       | The hash of the block to propose. This can be derived from the block header.                               | 1            |

Reviewer 2: > I'm thinking that maybe it's worth including the updated hash in the response call so that Tendermint can run a sanity check on the returned values.
Do you think it's worth it?

I think it is not worth it. I don't think we should have applications calculating this piece of data if the only thing that can happen is that they get it wrong and we panic. It just doesn't seem like something that they need to really be concerned with from my perspective. If it's the same one from BeginBlock then it is the header hash. Fwiw, candidate state can just as easily be indexed by height and round, so application developers don't really require a way to make sure the two hashes match.

>     | hash                    | bytes                                       | The hash of the block to propose. This can be derived from the block header.                               | 1            |

Author: Thanks for the info. I had got it wrong then. I'll update the text to clarify it's the header's hash

    | header                  | [Header](../core/data_structures.md#header) | The header of the block to propose.                                                                        | 2            |
    | last_commit_info        | [LastCommitInfo](#lastcommitinfo)           | Info about the last commit, including the round, the validator list, and which ones signed the last block. | 3            |

>     | byzantine_validators    | repeated [Evidence](#evidence)              | List of evidence of validators that acted maliciously.                                                     | 4            |

Reviewer 2: Is there a strong reason to pass this up to the application at this step? I'm not sure I see the use for it being passed up. I don't really think the application should be given the easy ability to modify this list and I'm not sure I see what it's supposed to do with it.

Reviewer 2: This makes total sense to me now, thanks!

>     | tx                      | repeated bytes                              | Preliminary list of transactions that have been picked as part of the block to propose.                    | 5            |

Reviewer 2: Oh, I see now. There are two possible actions that can be taken by an application. The application can mark the Tx as 'remove', in which case the Tx will be removed from the mempool. Alternatively, the application can _remove_ the application from the Tx list, in which case it stays in the mempool but is not in the proposal. 

I think we should stick with one format for modification of the list of Tx; namely, I think that if an application wants to alter a Tx, it should have to return the annotated Tx with the enum value relaying what behavior it wants. If you want it to not appear in _this specific block_, then marking it as 'UNPROPOSE' or 'REDACT' or similar seems a bit clearer. This is somewhat of a minor point either way. This would also allow the list of Tx returned to tendermint to be just a list of transformations or modifications. Absence from the list of modifications means 'no changes to this Tx' instead of 'do not propose at this time'.

>     | tx                      | repeated bytes                              | Preliminary list of transactions that have been picked as part of the block to propose.                    | 5            |

Author: Ok. let me see if I understood what you are proposing. The changes would be:
* Delete the "UNMODIFIED" value from the `TxAction` enum.
* Added a "DELAYED" or "UNPROPOSE" value to the `TxAction` enum
If a transaction is to appear unmodified, we simply remove it from the list. If it is to be removed from the proposal but kept in the mempool, we mark it as "UNPROPOSE".
If my understanding is correct, I see an issue with this solution. Since we are to support transaction reordering, I don't clearly see how we could combine reordered, modified, and unmodified transactions, since we would remove the unmodified ones from the list.

>     | tx                      | repeated bytes                              | Preliminary list of transactions that have been picked as part of the block to propose.                    | 5            |

Reviewer 2: That's a great point and as a result I think that my suggestion is incompatible with the requirement of allowing Tx to be reordered.

    | validator_updates       | repeated [ValidatorUpdate](#validatorupdate)     | Changes to validator set (set voting power to 0 to remove).                                 | 4            |
    | consensus_param_updates | [ConsensusParams](#consensusparams)              | Changes to consensus-critical gas, size, and other parameters.                              | 5            |

* **Usage**:
    * The Application can modify the parameters received in `RequestPrepareProposal` before sending
    them in `ResponsePrepareProposal`. In that case, `ResponsePrepareProposal.modified` is set to true.
    * In same-block execution mode, the Application can (and should) modify `ResponsePrepareProposal.data`,
      `ResponsePrepareProposal.validator_updates`, and `ResponsePrepareProposal.consensus_param_updates`.
    * In next-block execution mode, the Application can only modify `ResponsePrepareProposal.tx`, Tendermint
      will ignore any modification to the other fields.
    * If `ResponsePrepareProposal.modified` is false, then Tendermint should ignore the rest of
      parameters in `ResponsePrepareProposal`.

>     * As a sanity check, Tendermint will check the returned parameters for validity if the Application modified them.

Reviewer 2: These seem to capture the two main possibilities. My preference is for the second option, since we don't really have a consensus failure if a proposer behaves incorrectly, but I think we should get others to way in on this as well.

>     * As a sanity check, Tendermint will check the returned parameters for validity if the Application modified them.

Author: Ok, I'll add this as a TODO in the text so other reviewers can voice their opinion

          * the sum of `GasWanted` field of transactions $\in C$ is less than `consensusParams.block.max_gas`
      * _p_'s Tendermint creates a block header.
      * _p_'s Tendermint uses the newly generated block as block _v_
2. _p_'s Tendermint calls `RequestPrepareProposal` with block _v_.
   The call is synchronous: Tendermint's execution will block until the Application returns from the call.
3. The Application checks the block (header, transactions, commit info, evidences). Besides,
   * in "same-block execution" mode, the Application can (and should) provide `ResponsePrepareProposal.data`,
     `ResponsePrepareProposal.validator_updates`, or
     `ResponsePrepareProposal.consensus_param_updates`.
   * in "next-block execution" mode, _p_'s Tendermint will ignore the values for `ResponsePrepareProposal.data`,
     `ResponsePrepareProposal.validator_updates`, and `ResponsePrepareProposal.consensus_param_updates`.
   * in both modes, the Application can manipulate transactions
      * leave transaction untouched
      * add new transactions (not previously in the mempool)
      * mark transactions is invalid (remove)

>       * remove transactions (effectively _delaying_ them)

Reviewer 2: I see this now. You can mark it as `REMOVE` or remove it from the list. I think that the text here is a little ambiguous and that my preference is better captured in my comment above: We should make the `modifications` a list of just `modifications` and if a tx is absent from the list of modifications, then it will be proposed _as is_ by tendermint.

>       * remove transactions (effectively _delaying_ them)

Author: See my replay above about reordering.

      * modify transactions (e.g. aggregate them)
      * reorder transactions
   * **TODO**: include logic of `VerifyVoteExtension` here?
4. If the block is modified, the Application sets `ResponsePrepareProposal.modified` to true,
   and includes the modified block in the return parameters (see the rules in section _Usage_).
   The Application returns from the call.
5. _p_'s Tendermint uses the (possibly) modified block as _p_'s proposal in round _r_, height _h_.

### ProcessProposal

#### Parameters and Types

* **Request**:

    | Name                 | Type                                        | Description                                                                                                 | Field Number |
    |----------------------|---------------------------------------------|-------------------------------------------------------------------------------------------------------------|--------------|
    | hash                 | bytes                                       | The proposed block's hash. This can be derived from the block header.                                       | 1            |

>     | header               | [Header](../core/data_structures.md#header) | The proposed block's header.                                                                                | 2            |

Reviewer 2: I'm not sure we should reuse the same data structures here as those used in consensus. This creates a really tight coupling between our external API and the ABCI API.

Reviewer 2: Ah, I see. I was more concerned about which types _at the code level_ we were going to use here. Having them contain the same data is totally fine, but I think having them be the same protos may be a bit risky but that is something we can discuss as we begin the implementation.

    | accept | bool | If false, instruct Tendermint to prevote $\bot$ for this proposal | 1            |

* **Usage**:
    * Contains a full proposed block.
        * The parameters and types of `RequestProcessProposal` are the same as `RequestFinalizeBlock`
          (**TODO**: We'll see...).
        * In "same-block execution" mode, the Application will fully execute the block
          as though it was handling `RequestFinalizeBlock`.
          However, any resulting state changes must be kept as _canditade state_,
          and the Application should be ready to
          backtrack/discard it in case the decided block is different.
    * If `ResponseProcessProposal.accept` is true, Tendermint should prevote according to
      the normal procedure; else, it should prevote $\bot$.
    * The [Header](../core/data_structures.md#header) contains a boolan `problem` that indicates
      if the block is the result of an Application rejection (i.e., $\bot$). This allows the
      users to distinguish empty blocks because of lack of activity and empty blocks representing
      $\bot$, i.e., a problem in the code of `PrepareProposal` and/or `ProcessProposal`.

>     * The implementation of `ProcessProposal` MUST be deterministic. Moreover, the value of

Reviewer 2: Should we bypass this logic if `validValue != nil`? Perhaps an mild optimization if this method must be deterministic.

* **Usage**:
    * If `ResponseVerifyVoteExtension.accept` is false, Tendermint will flag the extension as _invalid_ but the
      validity of the received Precommit message will not be influenced by this.
    * The implementation of `VerifyVoteExtension` MUST be deterministic. Moreover, the value of
      `VerifyVoteExtension.accept` MUST *exclusively* depend on the parameters passed in
      the call to `RequestVerifyVoteExtension`, and the last committed Application state.

#### When does Tendermint call it?

When a validator _p_ is in Tendermint consensus round _r_, height _h_, state _prevote_ (**TODO** discuss: I think I must remove the state
from this condition, but not sure), and _p_ receives a Precommit message for round _r_, height _h_ from _q_:

1. _p_'s Tendermint calls `RequestVerifyVoteExtension`.
2. The Application returns _accept_ or _reject_ via `ResponseVerifyVoteExtension.accept`.
3. If the Application returns _reject_, _p_'s Tendermint flag the extension as _invalid_.
4. _p_'s Tendermint keeps all received votes, together with their corresponding vote extensions
   (and possibly _invalid_ flags) in its internal data structures. These will be used to:

>    * calculate field _LastCommitHash_ in the header of the block proposed for height _h + 1_

Reviewer 2: Maybe a minor point, but only validators need to store these, right? Full nodes, which do not produce blocks, will just check that the `last_commit` included in the block matches the header data.

    | log        | string                    | The output of the application's logger. **May be non-deterministic.** | 3            |
    | info       | string                    | Additional information. **May be non-deterministic.**                 | 4            |
    | gas_wanted | int64                     | Amount of gas requested for transaction.                              | 5            |
    | gas_used   | int64                     | Amount of gas consumed by transaction.                                | 6            |
    | tx_events  | repeated [Event](#events) | Type & Key-Value events for indexing transactions (e.g. by account).  | 7            |
    | codespace  | string                    | Namespace for the `code`.                                             | 8            |

### TxAction

```proto
  enum TxAction {
    UNKNOWN       = 0;  // Unknown action
    UNMODIFIED    = 1;  // The Application did not modify this transaction. Ignore old_hash field
    ADDED         = 2;  // The Application added this transaction. Ignore old_hash field

>     REMOVED       = 3;  // The Application wants this transaction removed from the proposal and the mempool. Ignore old_hash field

Author: Two points:
* We revisited this enum at the beginning of the week but I haven't had the time to commit it (sorry for the inconvenience). You may want to take a look, I'll commit it together with the changes I'm making to address your comments
* Anyway, your concern here is common to both versions, so here's my explanation
    * If the App wants the transaction delayed, it removes it completely from the returned list. Tendermint won't see it, so won't do anything with it: it'll stay in the mempool, and it won't be in the proposal.
    * If the App wants to signal that the transaction doesn't make sense (any longer), it leaves it in the list, and sets `action` to "REMOVED". Tendermint treats this the same way as CheckTx returning false

>     REMOVED       = 3;  // The Application wants this transaction removed from the proposal and the mempool. Ignore old_hash field

Reviewer 2: Thanks for clarifying, that clears up the ambiguity completely.

    |-------------------------|---------------------------------------------|------------------------------------------------------------------------------------------------------------------|--------------|
    | hash                    | bytes                                       | The block header's hash of the block to propose. Present for convenience (can be derived from the block header). | 1            |
    | header                  | [Header](../core/data_structures.md#header) | The header of the block to propose.                                                                              | 2            |
    | tx                      | repeated bytes                              | Preliminary list of transactions that have been picked as part of the block to propose.                          | 3            |
    | byzantine_validators    | repeated [Evidence](#evidence)              | List of evidence of validators that acted maliciously.                                                           | 4            |

>     | last_commit_info        | [LastCommitInfo](#lastcommitinfo)           | Info about the last commit, including the round, the validator list, and which ones signed the last block.       | 5            |

Author: Yes... need to add modifications to those data structures when we move them over

>     | last_commit_info        | [LastCommitInfo](#lastcommitinfo)           | Info about the last commit, including the round, the validator list, and which ones signed the last block.       | 5            |

Author: Discussed in person: Will add TODO here on what we need to modify

    | data                    | bytes                                            | The Merkle root hash of the application state.                                              | 3            |
    | validator_updates       | repeated [ValidatorUpdate](#validatorupdate)     | Changes to validator set (set voting power to 0 to remove).                                 | 4            |
    | consensus_param_updates | [ConsensusParams](#consensusparams)              | Changes to consensus-critical gas, size, and other parameters.                              | 5            |

* **Usage**:
    * Contains a preliminary block to be proposed, which the Application can modify.
    * The parameters and types of `RequestPrepareProposal` are the same as `RequestProcessProposal`
      and `RequestFinalizeBlock`.
    * The header contains the height, timestamp, and more - it exactly matches the
      Tendermint block header.
    * The Application can modify the parameters received in `RequestPrepareProposal` before sending
    them in `ResponsePrepareProposal`. In that case, `ResponsePrepareProposal.modified` is set to true.

>     * In same-block execution mode, the Application can (and should) modify `ResponsePrepareProposal.data`,

Author: Yes. i need to add a paragraph somewhere at the beginning. Part of the discussion on  how to split

>     * In same-block execution mode, the Application can (and should) modify `ResponsePrepareProposal.data`,

Author: Discussed in person: Will introduce them in the README.md
Then, will "tag" each "App mode" with the exec mode it enables (in app modes section)

      will ignore any modification to the other fields.
    * If `ResponsePrepareProposal.modified` is false, then Tendermint will ignore the rest of
      parameters in `ResponsePrepareProposal`.
    * As a result of executing the block to propose, the Application may produce header events or transaction events.
      The Application must keep those events until a block is decided and then pass them on to Tendermint via
      `ResponseFinalizeBlock`.
    * Likewise, the Application must keep all responses to executing transactions until it can call `ResponseFinalizeBlock`.
    * The Application can change the transaction list via `ResponsePrepareProposal.tx`.
      See [TransactionRecord](#transactionrecord) for further information on how to use it. Some notes:
        * To remove a transaction from the proposed block the Application _marks_ the transaction as
          "REMOVE". It does not remove it from the list.
        * Removing a transaction from the list means it is too early to propose that transaction,
          so it will be excluded from the proposal but will stay in the mempool for later proposals.

>           The Application should be extra-careful with this feature to avoid leaks in the mempool.

Author: The TX could stay forever in the mempool if the App misuses this

>           The Application should be extra-careful with this feature to avoid leaks in the mempool.

Author: Discussed in person: Rephare leak as something like "a TX can stay in the mempool forever if the App doesn't use this correctly"

    | accept | bool | If false, the received block failed verification | 1            |

* **Usage**:
    * Contains a full proposed block.
        * The parameters and types of `RequestProcessProposal` are the same as `RequestPrepareProposal`
          and `RequestFinalizeBlock`.
        * In "same-block execution" mode, the Application will fully execute the block
          as though it was handling `RequestFinalizeBlock`.
          However, any resulting state changes must be kept as _canditade state_,
          and the Application should be ready to
          backtrack/discard it in case the decided block is different.
    * The header contains the height, timestamp, and more - it exactly matches the
      Tendermint block header.
    * If `ResponseProcessProposal.accept` is _false_, Tendermint assumes the proposal received
      is not valid.
    * The implementation of `ProcessProposal` MUST be deterministic. Moreover, the value of
      `ResponseProcessProposal.accept` MUST *exclusively* depend on the parameters passed in
      the call to `RequestProcessProposal`, and the last committed Application state

>       (see [Properties](#properties) section below).

Author: Agreed

>       (see [Properties](#properties) section below).

Author: Discussed in person: Will reiterate in README.md when  introducing the functions

* **Usage**:
    * As `new_hashes` is a list, `TransactionRecord` allows to trace many-to-many modifications. Some examples:
        * Transaction $t1$ modified into $t2$ is represented with these records
            * $t2$ "ADDED"
            * $t1$ "REMOVED"; `new_hashes` contains [$id(t2)$]
        * Transaction $t1$ modified into $t2$ and $t3$ is represented with these `TransactionRecord` records
            * $t2$ "ADDED"
            * $t3$ "ADDED"
            * $t1$ "REMOVED"; `new_hashes` contains [$id(t2)$, $id(t3)$]
        * Transactions $t1$ and $t2$ aggregated into $t3$ is represented with these `TransactionRecord` records
            * $t3$ "ADDED"
            * $t1$ "REMOVED"; `new_hashes` contains [$id(t3)$]
            * $t2$ "REMOVED"; `new_hashes` contains [$id(t3)$]
        * Transactions $t1$ and $t2$ combined into $t3$ and $t4$ is represented with these `TransactionRecord` records
            * $t3$ "ADDED"
            * $t4$ "ADDED"
            * $t1$ "REMOVED" and `new_hashes` containing [$id(t3)$, $id(t4)$]
            * $t2$ "REMOVED" and `new_hashes` containing [$id(t3)$, $id(t4)$]

## Properties

> ### What Tendermint expects from the Application

Author: Let's discuss online on this

> ### What Tendermint expects from the Application

Author: Discussed in person: "Title of page (i.e., section) will be Requirements", and individual properties also rename to requirements

Let $p$ and $q$ be two different correct proposers in rounds $r_p$ and $r_q$ respectively, in height $h$.
Let $s_{p,h-1}$ be $p$'s Application's state committed for height $h-1$.

> Let $v_p$ (resp. $v_q$) be the block that $p$'s (resp. $q$'s) Tendermint passes on to the Application via `RequestPrepareProposal`

Author: Good idea. Will adapt

as proposer of round $r_p$ (resp $r_q$), height $h$.

> Let $v'_p$ (resp. $v'_q$) the possibly modified block $p$'s (resp. $q$'s) Application returns via `ResponsePrepareProposal` to Tendermint.

Author: Yep

> Let $v'_p$ (resp. $v'_q$) the possibly modified block $p$'s (resp. $q$'s) Application returns via `ResponsePrepareProposal` to Tendermint.

Author: Discussed in person: agreed on "Prepared Proposal"

The value proposed by $p$ can differ in two different rounds where $p$ is the proposer.

* Property 1 [`PrepareProposal`, header-changes] When the blockchain is in same-block execution mode,

>   $p$'s Application can change the following parameters in `ResponsePrepareProposal`:

Author: Yes :-)

>   $p$'s Application can change the following parameters in `ResponsePrepareProposal`:

Author: Discussed in person: I'll rephrase

  mode, $p$'s Application cannot make any changes to the following parameters in `ResponsePrepareProposal`:
  _AppHash_, _ConsensusParams_, _ValidatorUpdates_.

In practical terms, Property 2 implies that Tendermint will ignore those parameters in `ResponsePrepareProposal`.

* Property 3 [`PrepareProposal`, `ProcessProposal`, coherence]: For any two correct processes $p$ and $q$,
  if $q$'s Tendermint calls `RequestProcessProposal` on $v'_p$,
  $q$'s Application returns Accept in `ResponseProcessProposal`.

Property 3 makes sure that blocks proposed by correct processes _always_ pass the correct receiving process's
`ProcessProposal` check.
On the other hand, if there is a deterministic bug in `PrepareProposal` or `ProcessProposal` (or in both),
strictly speaking, this makes all processes that hit the bug byzantine. This is a problem in practice,
as very often validators are running the Application from the same codebase, so potentially _all_ would
likely hit the bug at the same time. This would result in most (or all) processes prevoting `nil`, with the
serious consequences on Tendermint's liveness that this entails.

> * Property 4 [`ProcessProposal`, determinism-1]: For any correct process $p$, and any arbitrary block $v'$,
>   if $p$'s Tendermint calls `RequestProcessProposal` on $v'$ at height $h$,
>   then $p$'s Application's acceptance or rejection exclusively depends on $v'$ and $s_{p,h-1}$.

Reviewer: ```suggestion
* Property 4 [`ProcessProposal`, determinism-1]: `ProcessProposal` is a (deterministic) function of the current state and the block that is about to be applied. In other words, for any correct process $p$, and any arbitrary block $v'$,
  if $p$'s Tendermint calls `RequestProcessProposal` on $v'$ at height $h$,
  then $p$'s Application's acceptance or rejection **exclusively** depends on $v'$ and $s_{p,h-1}$.
```

* Property 5 [`ProcessProposal`, determinism-2]: For any two correct processes $p$ and $q$, and any arbitrary block $v'$,
  if $p$'s (resp. $q$'s) Tendermint calls `RequestProcessProposal` on $v'$ at height $h$,
  then $p$'s Application accepts $v'$ if and only if $q$'s Application accepts $v'$.
  Note that this Property follows from Property 4 and the Agreement property of consensus.

Properties 4 and 5 ensure that all correct processes will react in the same way to a proposed block, even
if the proposer is Byzantine. However, `ProcessProposal` may contain a bug that renders the
acceptance or rejection of the block non-deterministic, and therefore prevents processes hitting
the bug from fulfilling Properties 4 or 5 (effectively making those processes Byzantine).
In such a scenario, Tendermint's liveness cannot be guaranteed.
Again, this is a problem in practice if most validators are running the same software, as they are likely
to hit the bug at the same point. There is currently no clear solution to help with this situation, so
the Application designers/implementors must proceed very carefully with the logic/implementation

> of `ProcessProposal`. As a general rule `ProcessProposal` _should_ always accept the block.

Reviewer: Also here we could add a note that this is a target for testing and verification.

Reviewer: I guess this comment applies to all the properties ;-)

Author: Yes. And I also had an idea of a possible tool that would use these properties. Let's talk like

> According to the Tendermint algorithm, a correct process can only broadcast one precommit message in round $r$, height $h$.

Reviewer: ```suggestion
According to the Tendermint algorithm, a correct process can broadcast at most one precommit message in round $r$, height $h$.
```

  then $p$'s Application's acceptance or rejection exclusively depends on $e$, $w$ and $s_{p,h-1}$.

* Property 8 [`VerifyVoteExtension`, determinism-2]: For any two correct processes $p$ and $q$,
  and any arbitrary vote extension $e$, and any arbitrary block $w$,
  if $p$'s (resp. $q$'s) Tendermint calls `RequestVerifyVoteExtension` on $e$ and $w$ at height $h$,
  then $p$'s Application accepts $e$ if and only if $q$'s Application accepts $e$.
  Note that this property follows from Property 7 and the Agreement property of consensus.

Properties 7 and 8 ensure that the validation of vote extensions will be deterministic at all
correct processes.
Properties 7 and 8 protect against arbitrary vote extension data from Byzantine processes
similarly to Properties 4 and 5 and proposed blocks.
Properties 7 and 8 can be violated by a bug inducing non-determinism in `ExtendVote` or
`VerifyVoteExtension`. In this case liveness can be compromised.
Extra care should be put in the implementation of `ExtendVote` and `VerifyVoteExtension` and,
as a general rule, `VerifyVoteExtension` _should_ always accept the vote extension.

> * Property 9 [_all_, read-only]: $p$'s calls to `RequestPrepareProposal`, `RequestProcessProposal`,

Reviewer: Should we say "free of side effects" rather than "read-only"?

Author: Yep. Look reasonable

  `RequestExtendVote`, and `RequestVerifyVoteExtension` at height $h$ do not modify $s_{p,h-1}$.

* Property 10 [`ExtendVote`, `FinalizeBlock`, non-dependency]: for any correct process $p$,
and any vote extension $e$ that $q$ received at height $h$, the computation of
$s{p,h}$ does not depend on $e$.

The call to correct process $p$'s `RequestFinalizeBlock` at height $h$, with block $v_{p,h}$
passed as parameter, creates state $s_{p,h}$.
Additionally, $p$'s `FinalizeBlock` creates a set of transaction results $T_{p,h}$.

>**TODO** I have left out all the "events" as they don't have any impact in safety or liveness
>(same for consensus params, and validator set)

* Property 11 [`FinalizeBlock`, determinism-1]: For any correct process $p$,

>   the contents of $s_{p,h}$ exclusively depend on $s_{p,h}$ and $v_{p,h}$.

Reviewer: I guess this should be rephrased. I am not sure about "contents". Also there is a wrong index in "$s_{p,h}$ exclusively depend on $s_{p,h}$"

Author: Yes. Thanks for catching. Fix coming with the commit I am about to push
